syntax = "proto3";

package resemble.v1alpha1;

import "google/api/annotations.proto";

////////////////////////////////////////////////////////////////////////

message QueryRequest {
  // Fully-qualified name of a Resemble unary reader method, e.g.,
  // `my.package.Service.ReaderMethod`. Only unary reader methods are
  // supported.
  string method = 1;

  // Serialized request to be passed to the above method. The same
  // request is used each time we invoke the method.
  bytes request = 2;
}

message QueryResponse {
  // Serialized response from the method specified in the request.
  bytes response = 1;

  // Set of idempotency keys that caused mutations since the last time
  // that a response was returned.
  repeated string idempotency_keys = 2;
}

////////////////////////////////////////////////////////////////////////

service React {
  // Allows users to "watch" the response of a unary reader method for
  // changes. The current response of the method is sent when the
  // stream is opened, then a new response is sent whenever the
  // actor's state changes in a way that causes the method's response
  // to change.
  //
  // NOTE: the service name and actor ID is expected to be part of the
  // gRPC metadata, exactly like any other resemble requests.
  rpc Query(QueryRequest) returns (stream QueryResponse) {
    option (google.api.http) = {
      post: "/query"
    };
  }
}

////////////////////////////////////////////////////////////////////////
